---
title: ""
author: "Rodrigo Arcoverde"
date: '`r format(Sys.Date(), "%Y-%m-%d")`'
output: html_document
abstract: Short project description.
knit: (function(inputFile, encoding) {
          rmarkdown::render(inputFile,
                            encoding = encoding, 
                            output_file = paste0(
                              xfun::sans_ext(inputFile), '_', Sys.Date(), '.html'),
                                output_dir = "../results/lab_book/")})
---

```{r setup, include=FALSE}
knitr::opts_knit$set(echo = TRUE,
                     root.dir = getwd(),
                     fig.width = 6, fig.height = 5,
                     warning = FALSE, 
                     message = FALSE)

result.dir <- paste0("results/",Sys.Date(),"/")

## creates result.dir with date in if not existent
ifelse(isFALSE(dir.exists(paste0("../",result.dir))), dir.create(paste0("../",result.dir),recursive = TRUE),"Result directory for today exists already!")
options(stringsAsFactors = FALSE) 
```

## Needed libraries
```{r libraries, message=FALSE}
library(ggpubr)
library(RColorBrewer)
library(gplots)
library(reshape2)
library(dplyr)
library(ggplot2)
library(Biostrings)
library(cowplot)
library(ggsci)
library(ggtree)
library(treeio)
library(vegan)
## creating needed custom functions
# function to create fasta file from data frame
.df_to_fasta <- function(sequences_column, sequence_name){
  if(length(sequences_column) != length(sequence_name)){
    print("Sequences columng does not have the same length as sequences name")
  }
  else {
    str <- AAStringSet(sequences_column)
    names(str) <- sequence_name
  }
  return(str)
}

colors <- c("group_1" = "#FF2600", "group_2" = "#008F00", "group_3" = "#011993")
```

## Load data
```{r message=FALSE, warning=FALSE}
rabies_seq <- read.csv("../data/rabies_sequences/rabies_VH-sequence-database_AC.csv") %>%
  mutate(VDJ_aa = sub("\\*","",VDJ_aa))
filt_1 <- read.table("../results/2021-11-01/sequence_QC_pipeline/group_1/igdiscover_run/final/filtered.tab.gz", header = TRUE, sep = "\t") %>% mutate(VDJ_aa = sub("\\*","",VDJ_aa))
filt_2 <- read.table("../results/2021-11-01/sequence_QC_pipeline/group_2/igdiscover_run/final/filtered.tab.gz", header = TRUE, sep = "\t") %>% mutate(VDJ_aa = sub("\\*","",VDJ_aa))
filt_3 <- read.table("../results/2021-11-01/sequence_QC_pipeline/group_3/igdiscover_run/final/filtered.tab.gz", header = TRUE, sep = "\t") %>% mutate(VDJ_aa = sub("\\*","",VDJ_aa))

ls <- list(filt_1,filt_2,filt_3)
names(ls) <- c("group_1","group_2","group_3")
merged_df <- data.table::rbindlist(ls,idcol = TRUE)
#write.table(merged_df, paste0("../", result.dir, "filtered_combined.csv"),row.names = FALSE, quote = FALSE, sep = ",")


rabies_1 <- merged_df %>% filter(.id == "group_1")
rabies_2 <- merged_df %>% filter(.id == "group_2")
rabies_3 <- merged_df %>% filter(.id == "group_3")

ls_grp <- list(rabies_1,rabies_2, rabies_3)

## saving each group as fasta files for alignment
for(i in 1:length(ls_grp)){
  fasta <- .df_to_fasta(sequences_column = ls_grp[[i]][["VDJ_aa"]], sequence_name = ls_grp[[i]][["name"]])
  writeXStringSet(fasta, paste0("../", result.dir, "aa_group_", i,".fasta"))
}

#all_seq_nt <- .df_to_fasta(sequences_column = rabies_seq$VDJ_nt, sequence_name = rabies_seq$Sequence.ID)
all_seq_aa <- .df_to_fasta(sequences_column = merged_df$VDJ_aa, sequence_name = merged_df$name)


#writeXStringSet(all_seq_nt, paste0("../", result.dir, "nt_all",".fasta"))
writeXStringSet(all_seq_aa, paste0("../", result.dir, "aa_all",".fasta"))

## Proceeed to igdiscover to run sequences using Gunilla Macaca Mullata as starting database

```

## Comparing SHM 

### Normality distribution

```{r}
## Quantile-quantile plots to evaluate theoretical vs observed quantile data
merged_df %>%
  ggplot(aes(sample = V_SHM, color = .id)) +
  stat_qq() + stat_qq_line() +
  facet_grid(~.id) +
  theme_bw()+
  theme(legend.position = "none") +
  scale_color_manual(values = colors) 

shapiro.test(merged_df[merged_df$.id == "group_1", V_SHM])
shapiro.test(merged_df[merged_df$.id == "group_2", V_SHM])
shapiro.test(merged_df[merged_df$.id == "group_3", V_SHM])

```

### Somatic hypermutations between groups

```{r}
## function to add summary to box plot
stat_box_data <- function(y, upper_limit = max(iris$Sepal.Length) * 1.15) {
  return( 
    data.frame(
      y = 0.95 * upper_limit,
      label = paste('count =', length(y), '\n',
                    'mean =', round(mean(y), 1), '\n')
    )
  )
}

## filtered tabs from igdiscover output processing
filtered <- read.table("../results/2021-10-20/nt_all/final/filtered.tab", sep = "\t", header = T) %>%
  mutate(Sequence.ID = name)

# joining information from the two different datasets
merged_info <- left_join(filtered,rabies_seq[1:6]) %>%
        mutate(Group = plyr::mapvalues(x = Color.coding, from = c(1,2,3), to = c("Group 1", "Group 2", "Group 3")))

#comparing means,showing that the comparison using BH correction is significant by kruskal.test
kruskal.test(V_SHM ~ .id,  data = merged_df)
#doing multiple comparisons afterwards, correcting by BH aka False Discovery Rate p.adjust.method, comparing all to the reference group. Both significant after adjustments

compare_means(V_SHM ~ .id,  data = merged_df, ref.group = "group_3", p.adjust.method = "BH")

## plottings
merged_df %>% 
  ggplot(aes(.id, V_SHM, color= .id)) +
  geom_violin(draw_quantiles = c(0.25, 0.5, 0.75)) +
  geom_point() +
  scale_color_manual(values = colors)+
  stat_summary(
    fun.data = stat_box_data, 
    geom = "text", 
    hjust =-.1,
    vjust = 5.5) + 
 theme_cowplot()+ 
  #stat_compare_means(method = "t.test", label.y = 15)+        # Add global annova p-value
 stat_compare_means(method = "kruskal.test",
                     label.y = 15)+
  labs(x = "",y = "% of VH SHM") +
  theme(legend.position = "none")

ggsave(paste0("../", result.dir, "_v-shm.eps"),
       width = 17, height = 15, 
       units = "cm", limitsize = FALSE) 
```

## V-J pairing

### First heatmap

Plotting only counts, not accounting for number of sequences sequenced. Heatmaps not wrapped together, thus with different symm breaks scales.

```{r}
database_v <- Biostrings::readDNAStringSet("../data/kimdb_gkhlab_macaca-mulatta/V.fasta")
database_j <- Biostrings::readDNAStringSet("../data/kimdb_gkhlab_macaca-mulatta/J.fasta")

names_v <- unique(sub("\\*.*| ","",names(database_v)))
names_j <- unique(sub("\\*.*| ","",names(database_j)))
empty_matrix <- matrix(nrow = length(names_j), ncol = length(names_v))
colnames(empty_matrix) <- names_v
rownames(empty_matrix) <- names_j

for(i in c("Group 1", "Group 2", "Group 3")){
  
  df_heatmap <- merged_info %>% 
    filter(Group == i) %>%
      mutate(J_gene = sub("\\*.*| ","",J_gene),
           V_gene = sub("\\*.*| ","", V_gene)) %>%
    dplyr::count(V_gene, J_gene, Group) %>%
    mutate(freq = n) %>%
    #filter(freq > 1) %>%
    select(-c("Group", "n")) %>%
    acast(J_gene~V_gene) 

    # filling up empty matrix 
  cols <- colnames(empty_matrix)[colnames(empty_matrix) %in% colnames(df_heatmap)]
  rows <- rownames(empty_matrix)[rownames(empty_matrix) %in% rownames(df_heatmap)]

  empty_matrix[rows, cols] <- df_heatmap[rows, cols]
  heatmap_data <- empty_matrix

  heatmap_data[is.na(heatmap_data)] <- 0

  setEPS()
  postscript(paste0("../", result.dir, Sys.Date(), "_", i,"V-J_pairing.eps"))


  heatmap.2(heatmap_data, 
           lwid =  c(0.5,2),
           lhei = c(1,2),
           density.info = "none",
           trace = "none",
           cexRow = .8,
           cexCol = .3,
           col = viridis::mako(99, direction = -1, begin = 0),
           dendrogram = "none",
           symkey = FALSE,
           margins=c(5,4),
           breaks = 100) 
dev.off()
}
```

### Second heatmap

Account for number of sequences (V-J pairing percentage %), plots wrapped together for easier comprehession, V genes maintained. 

```{r}


empty_matrix <- matrix(nrow = length(names_j), ncol = length(names_v))
colnames(empty_matrix) <- names_v
rownames(empty_matrix) <- names_j


df_heatmap <- merged_info %>% 
     mutate(J_gene = sub("\\*.*| ","",J_gene),
           V_gene = sub("\\*.*| ","", V_gene)) %>%
    dplyr::count(V_gene, J_gene, Group) %>%
    mutate(freq = n/sum(n)*100)


expand.grid(names_v,names_j, c("Group 1", "Group 2", "Group 3")) %>%
  as.data.frame() %>%
  dplyr::rename(V_gene = Var1, J_gene = Var2, Group = Var3) %>%
  full_join(df_heatmap) %>%
  tidyr::replace_na(list(freq = 0, n = 0)) %>%
  mutate(V_gene_family = substr(V_gene, 1,5)) %>%
  ggplot(aes(V_gene, J_gene, fill = freq)) +
  geom_tile(color = "black") +
  scale_fill_viridis_c(option = "viridis", direction = 1)+
  theme(axis.text.x = element_text(angle = 90, size = 7, hjust = 1, vjust = 0.5, face = "bold", colour = "black"),
        axis.text.y = element_text(face = "bold", colour = "black"),
        strip.background = element_rect(color="black", fill="grey90", size=.8, linetype="solid"),
        strip.text = element_text(face = "bold"),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        legend.position =  "top")+
  labs(x = "IGHV genes", y = "IGHJ genes",  fill = "V-J pairing\nfrequency\n(%)") +
  facet_wrap(~Group, ncol = 1)

ggsave(paste0("../", result.dir, "V-J_pairing.pdf"), 
       dpi = 300, width = 29,height = 15, units = "cm" ,
       limitsize = FALSE) 

```

## HCDR3 length comparison

```{r}
merged_df %>%
  ggplot(aes(nchar(CDR3_aa), fill = .id)) +
  geom_histogram(color = "black", bins = 19) +
   theme_cowplot() +
    labs(x = "CDR3 aa length", y = "Count", title = "") +
  scale_fill_manual(values = colors)+
  facet_grid(~.id) +
  theme(legend.position = "none",
        plot.title = element_text(hjust = 0.5),
        strip.background =element_rect(fill="grey95")) 

ggsave(paste0("../", result.dir, "_cdr3-length.eps"),
       width = 17, height = 15, 
       units = "cm", limitsize = FALSE) 
```

## Plotting trees 

```{r}

ls <- list.files("../results/2021-11-01/trees", full.names = T)
trees <- lapply(ls, treeio::read.tree)
names(trees) <- lapply(ls, function (x) strsplit(x, "/")[[1]][[5]])



colors_tree <- c("Ref. Site I" = "#FF40FF", "Ref. Site III" = "#00E9EB",
                 "Group 1" = "#FF2600", "Group 2" = "#008F00", "Group 3" = "#011993")

colors_tree_2 <- c("Ref. Site I" = "#FF40FF", "Ref. Site III" = "#00E9EB",
                 "High SHM" = "#000000", 
                 "Medium SHM" = "#898989",
                 "Low SHM" = "#C8C8C8")

rabies_df <- rabies_seq %>%
  mutate(.id = ifelse(grepl("Ref", Group), Group, paste0("Group ", Group)),
         name = Sequence.ID,
         site = ifelse(grepl("Ref", Group), Group, 
                       ifelse(grepl("Med", mAb), "Medium SHM",
                              ifelse(grepl("High", mAb), "High SHM",
                              ifelse(grepl("Low", mAb), "Low SHM", ""))))) %>%
  select(name,.id, mAb, site) %>%
  as.data.frame() %>% 
  left_join(merged_info, by = "name") %>%
  mutate(V_gene_family = substr(V_gene, 1,5)) %>%
  select(name, .id, mAB, site, V_gene_family)
  
rabies_seq[!rabies_seq$Sequence.ID %in% merged_info$name,]
  
rabies_seq$Sequence.ID %in% merged_info$name

rownames(rabies_df) <- rabies_df[,"name"]

for(i in seq_along(trees)){
  tree <- ggtree(trees[[i]]) %<+% rabies_df +
    geom_tiplab(aes(label = mAb), align = TRUE) + 
    geom_treescale()
  
  tree <- gheatmap(tree, rabies_df[".id"], width = .1, offset = .1) +
    scale_fill_manual(values = colors_tree)
  tree
  
  ggsave(plot = tree, filename = paste0("../", result.dir,Sys.Date(),names(trees)[[i]], ".pdf"), width = 15, height = 29)

  # to create slanted trees as well
  
  tree_2 <- ggtree(trees[[i]], layout = "equal_angle") %<+% rabies_df +
    geom_tiplab(aes(label = mAb), hjust = -.1) + 
    geom_tippoint(aes(color = site), size = 4) +
    scale_color_manual(values = colors_tree_2, na.value = "#ffffff00")+
    geom_treescale(width = .2) +
    labs(color = "") +
    theme_tree(text=element_text(family = "Helvetica"))
  tree_2
  ggsave(plot = tree_2, filename = paste0("../", result.dir,Sys.Date(),names(trees)[[i]], "slanted_tree.pdf"),device = "pdf", width = 20, height = 20)
}


```
## Clonotype diversity and richness

### Chao1 and ACE: species richness 

Diversity estimation using ```vegan``` package

```{r warning=FALSE, message=FALSE}


chaox100 <- function(x){
  replicate(100, {
  subsample <-  rrarefy(x,min(rowSums(x)))
  chao <- estimateR(subsample)
  return(chao)
})}


clone_size <- t(as.matrix(read.csv("../data/clonotypes_summary/clonotypes_size.csv")))
clone_size[is.na(clone_size)] <- 0

chao_clones <- chaox100(clone_size)
chao_clones_list <- lapply(asplit(chao_clones,1),t)

s.obs_clone <- as.data.frame((chao_clones_list$S.obs))
s.chao_clone  <- as.data.frame((chao_clones_list$S.chao1))
s.ace_clone <- as.data.frame((chao_clones_list$S.ACE))
chao <- estimateR(clone_size)

colors <- c("Group 1" = "#FF2600", "Group 2" = "#008F00", "Group 3"= "#011993")

g1 <- melt(s.obs_clone) %>%
  mutate(variable = sub(pattern = "group_", replacement = "Group ", x = variable)) %>%
  ggplot(aes(variable,value, fill = variable)) +
  geom_boxplot() +
  #stat_compare_means(label = "p.signif", method = "wilcox.test", ref.group = "group_3",
   #                  label.y = 2500, p.adjust = "bonferroni") +
  theme_cowplot(font_size = 12,line_size = 2, rel_small = 2, rel_large = 2,rel_tiny = 2) +
  scale_fill_manual(values = colors) +
  labs(y = "Clonotypes observed", x ="") +
  theme(legend.position = "none",
        axis.title.y = element_text(face = 2, size = 20))
g1
ggsave(g1,filename = paste0("../", result.dir,"observed_species.pdf"),device = "pdf", width = 10, height = 10)

g2 <- melt(s.chao_clone) %>%
  mutate(variable = sub(pattern = "group_", replacement = "Group ", x = variable)) %>%
  ggplot(aes(variable,value, fill = variable)) +
  geom_boxplot() +
  #stat_compare_means(label = "p.signif", method = "wilcox.test", ref.group = "group_3",
   #                  label.y = 2500, p.adjust = "bonferroni") +
  theme_cowplot(font_size = 12,line_size = 2, rel_small = 2, rel_large = 2,rel_tiny = 2) +
  scale_fill_manual(values = colors) +
  labs(y = "Clonotype richness estimator (Chao 1)", x ="") +
  theme(legend.position = "none",
        axis.title.y = element_text(face = 2, size = 20))
g2
ggsave(g2,filename = paste0("../", result.dir,"estimated_species.pdf"),device = "pdf", width = 10, height = 10)
```

## iNEXT (INterpolation and EXTrapolation): rarefraction and extrapolation of species diversity

```{r warning=FALSE, message=FALSE}
library(iNEXT)

rownames(clone_size) <- c("Group 1", "Group 2", "Group 3")

for(i in seq(3)-1){
  print(i)
inext_obj <- iNEXT(t(clone_size),q=c(i), datatype = "abundance")

  if(i == 0){
    ggiNEXT(inext_obj, type=1, se=TRUE, facet.var="none", color.var="site", grey=FALSE)  +
     theme_cowplot() +
      ggtitle("Species richness (Chao1)") +
      scale_color_manual(values = colors) +
      scale_fill_manual(values=colors)+
  theme(plot.title = element_text(hjust = 0.5, size = 28),
        axis.title = element_text(face = "bold", size = 24),
        axis.text = element_text(size = 20, face = "bold"),
        axis.line = element_line(size = 2),
        axis.ticks = element_line(size = 2),
        legend.text = element_text(face = "bold", size = 24))


      ggsave(filename = paste0("../", result.dir,"species_richness.pdf"),device = "pdf", width = 10, height = 8)
  }
  if(i == 1){
    ggiNEXT(inext_obj, type=1, se=TRUE, facet.var="none", color.var="site", grey=FALSE)  +
      theme_cowplot() +
      ggtitle("Shannon diversity") +
      scale_color_manual(values = colors) +
      scale_fill_manual(values=colors) +
  theme(plot.title = element_text(hjust = 0.5, size = 28),
        axis.title = element_text(face = "bold", size = 24),
        axis.text = element_text(size = 20, face = "bold"),
        axis.line = element_line(size = 2),
        axis.ticks = element_line(size = 2),
        legend.text = element_text(face = "bold", size = 24))


      ggsave(filename = paste0("../", result.dir,"shannon_diversity.pdf"),device = "pdf", width = 10, height = 8)
  }
  if(i == 2){
    ggiNEXT(inext_obj, type=1, se=TRUE, facet.var="none", color.var="site", grey=FALSE)  +
      theme_cowplot() +
      ggtitle("Simpson diversity") +
      scale_color_manual(values = colors) +
      scale_fill_manual(values=colors)+
  theme(plot.title = element_text(hjust = 0.5, size = 28),
        axis.title = element_text(face = "bold", size = 24),
        axis.text = element_text(size = 20, face = "bold"),
        axis.line = element_line(size = 2),
        axis.ticks = element_line(size = 2),
        legend.text = element_text(face = "bold", size = 24))


      ggsave(filename = paste0("../", result.dir,"simpson_diversity.pdf"),device = "pdf", width = 10, height = 8)
  }
}


inext_obj <- iNEXT(t(clone_size),q=c(0), datatype = "abundance")

ggiNEXT(inext_obj, type=2, se=TRUE, facet.var="none", color.var="site", grey=FALSE)  +
  theme_cowplot() +
  ggtitle("Sample coverage") +
  scale_color_manual(values = colors) +
  scale_fill_manual(values=colors) +
  theme(plot.title = element_text(hjust = 0.5, size = 28),
        axis.title = element_text(face = "bold", size = 24),
        axis.text = element_text(size = 20, face = "bold"),
        axis.line = element_line(size = 2),
        axis.ticks = element_line(size = 2),
        legend.text = element_text(face = "bold", size = 24))

ggsave(filename = paste0("../", result.dir,"sample_coverage.pdf"),device = "pdf", width = 10, height = 8)

```

## Take environment snapshot 
```{r}
renv::snapshot()
```


## SessionInfo
```{r}
sessionInfo()
```


